### DISCUSSION TITLE
Exclude services from nsclient.ini using wildcards?

### DISCUSSION THREAD

**Discussion by DrWebster** [mickem/nscp#828] [Q&A] [answered]
We have Nagios check the services on our Windows servers to ensure anything set to automatic start is running. Occasionally we'll need to exclude some services from the check, and this is easy enough to do in the affected server's nsclient.ini file as part of the alias_service stanza ("exclude=servicename"). Problem is, some applications such as Google Chrome install services with very specific names (like "GoogleUpdaterInternalService123.0.6288.0") that change every time the application is updated. Simply applying a wildcard to the service name (like "exclude=Google*") doesn't seem to work. Is there a way to exclude services using a wildcard or regex?

**Comment by jeanbruder** [comment]
Hi,

Same request here 'cause of Google updater too ... Have You found any solution @DrWebster ?

Regards,
Jean

**Reply by DrWebster** [reply]
Not yet. I'm probably going to have to open a thread on the Nagios forums to figure out a workaround.

**Comment by sjwk** [comment]
Yes, you can use 'filter'.  In my alias definition that uses check_service I have:
`filter="name not like 'clr_optimization' AND name not like 'GoogleUpdater'"`

I also have a bunch of exclude fields, it'll apply both.  The only caveat appears to be that you can only have one filter parameter, hence the AND above (having two filter parameters just ignored the second).  
More information: https://nsclient.org/docs/reference/windows/CheckSystem/#check_service

I build and deploy the settings file with ansible, and it's starting to get a bit unwieldy with the number of exclude parameters, I'm thinking of changing it and only checking for the services I actually need to be running, rather than all that are marked as Automatic...

**Reply by DrWebster** [reply]
Thanks. I think we must be doing something different with our service checks; when I add the "filter" statement like you have, Nagios returns this:

`Failed to validate filter see log for details: Critical expression is not valid:( name like 'filter' and state != 'name not like 'GoogleUpdater'' )`

Here's the beginning of our service check line from nsclient.ini (there are a few more "exclude" statements after it, and those all work fine):

`alias_service = checkServiceState CheckAll filter="name not like 'GoogleUpdater'" "exclude=ShellHWDetection"`

**Reply by sjwk** [reply, marked as answer]
What version are you using?  I'm running 0.5.2.41.
My alias is built using check_service rather than checkServiceState - it looks as though the latter is described in the docs as a legacy version of check_service and doesn't seem to have the Filter parameter.

I've created mine as:
`alias_service = check_service exclude="Net Driver HPZ12" exclude="Pml Driver HPZ12" exclude=stisvc exclude=swi_update_64 exclude="<snip another 20 or so exclude parameters here>" exclude="MdmIntegrationService" exclude="RCSServer" filter="name not like 'clr_optimization' AND name not like 'GoogleUpdater'"
`
As long as you're not running a really old version, try changing it to use check_service instead.  Also don't need the 'CheckAll' parameter with check_service that I can see in the docs.

**Reply by DrWebster** [reply]
Thank you, that did it!

**Reply by notorious-mso** [reply]
I had the same problem, now it's solved thanks to you,

**Comment by mickem** [comment]
Late confirmation of what @sjwk said, with the reasoning behind it:

- `exclude=` is an exact match on the service name, so `exclude=Google*` never matches anything. It is meant for a handful of fixed names.
- `filter=` is the general tool. It is the check's SQL-like filter language, and for strings it has `like` (substring), `not like`, `regexp` and `not regexp`:

  ```ini
  alias_service = check_service "filter=name not like 'GoogleUpdater' and name not like 'clr_optimization'"
  ```

  or, for the versioned Google names, a regex:

  ```ini
  alias_service = check_service "filter=name not regexp 'GoogleUpdater.*'"
  ```

  `filter=` replaces the default filter, which is empty for `check_service`, so nothing else changes: the default `warning`/`critical` (a service whose state does not match its start type) still apply to every service the filter lets through, and `exclude=` can still be combined with it for exact names.
- `checkServiceState` is the legacy 0.3/0.4 command; it does not understand `filter=`, which is where the "Failed to validate filter" came from. `check_service` is the current one and takes all the filter options.
